src/gui/properties/piecesbar.cpp

```diff
@@ -0,0 +1,332 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2016  Eugene Shalygin
+ * Copyright (C) 2006  Christophe Dumez
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "piecesbar.h"
+
+#include <QApplication>
+#include <QHelpEvent>
+#include <QPainter>
+#include <QTextStream>
+#include <QToolTip>
+
+#include "base/bittorrent/torrenthandle.h"
+#include "base/utils/misc.h"
+
+#define TYPESET_FILELIST_AS_TABLE // TODO remove this #ifdef before merging to master
+
+namespace
+{
+    class PieceIndexToImagePos
+    {
+    public:
+        PieceIndexToImagePos(const BitTorrent::TorrentInfo &torrentInfo, const QImage &image)
+            : m_bytesPerPixel {image.width() > 0 ? torrentInfo.totalSize() / image.width() : -1}
+            , m_pieceSize {torrentInfo.pieceLength()}
+        {
+        }
+
+        int imagePos(int pieceIndex) const
+        {
+            return m_bytesPerPixel < 0 ? 0 : (pieceIndex * m_pieceSize + m_pieceSize / 2) / m_bytesPerPixel;
```

Comment:
I suppose that you meant KiB, not MiB, for the sizes, then bytesPerPixel = 20.

Let's compute pixels per piece: pixelsPerPiece = totalPixels / numberOfPieces = totalPixels / ceil(torrrentSize / pieceSize); In this example numberOfPieces = 5 and pixelsPerPiece = 100. The method computes image positions for the center of the given piece. Exact number is 450, while computation gave 460. The error is 10/450 =  2.2%.

The estimation of the error is: `image.width() / torrentInfo.totalSize() * 100%` = 500 / 10240 *100% = 4.9%;

However, taking a torrent of size 10 MiB we get error value of approx 1920 / (10 * 2^20)*100% =  0.02% and thus a typical error will be 1920*0.02% = 0.384 px.

Should we want precise pointing, we need to compute two image positions: for the beginning of a piece a for the end. Perhaps that would make computations clearer, I don't know. Testing on you showed that I have to either put the approximation used into a comment or rewrite the code using two ends and, perhaps, floating point arithmetic.